server/market/market.go

```diff
@@ -3,30 +3,435 @@
+}
+
+// SubmitOrderAsync submits a new order for inclusion into the current epoch.
+// When submission is completed, an error value will be sent on the channel.
+// This is the asynchronous version of SubmitOrder.
+func (m *Market) SubmitOrderAsync(rec *orderRecord) <-chan error {
+	sig := newOrderUpdateSignal(rec)
+	m.orderRouter <- sig
+	return sig.errChan
+}
+
+// NewMarket creates a new Market for for the provided base and quote assets,
+// with an epoch cycling at given duration in seconds.
+func NewMarket(ctx context.Context, epochDurationSec int64, baseAsset, quoteAsset *asset.Asset,
+	storage db.DEXArchivist, authMgr swap.AuthManager) (*Market, error) {
+
+	if err := storage.LastErr(); err != nil {
+		return nil, err
+	}
+
+	lotSize := baseAsset.LotSize
+	mktInfo, err := types.NewMarketInfoFromSymbols(baseAsset.Symbol, quoteAsset.Symbol, lotSize)
+	if err != nil {
+		return nil, err
+	}
+
+	// Supplement the provided context with an cancel function. The Stop method
+	// may be used to stop the market, or the parent context itself may be
+	// canceled.
+	ctxInternal, cancel := context.WithCancel(ctx)
+
+	swapperCfg := &swap.Config{
+		Ctx: ctxInternal,
+		Assets: map[uint32]*asset.Asset{
+			mktInfo.Base:  baseAsset,
+			mktInfo.Quote: quoteAsset,
+		},
+		Storage:          storage,
+		AuthManager:      authMgr,
+		BroadcastTimeout: 10 * time.Second, // TODO: who sets this?
+	}
+
+	return &Market{
+		ctx:           ctxInternal,
+		cancel:        cancel,
+		running:       make(chan struct{}),
+		marketInfo:    mktInfo,
+		lotSize:       lotSize,
+		epochDuration: epochDurationSec,
+		orderRouter:   make(chan *orderUpdateSignal, 16),
+		book:          book.New(lotSize),
+		matcher:       matcher.New(),
+		swapper:       swap.NewSwapper(swapperCfg),
+		auth:          authMgr,
+		baseAsset:     baseAsset,
+		quoteAsset:    quoteAsset,
+		storage:       storage,
+	}, nil
+}
+
+// Start beings order processing.
+func (m *Market) Start() {
+	m.wg.Add(1)
+	go m.runEpochs()
+}
+
+// Stop begins the Market's shutdown. Use WaitForShutdown after Stop to wait for
+// shutdown to complete.
+func (m *Market) Stop() {
+	log.Infof("Market shutting down...")
+	m.cancel()
+	for _, s := range m.orderFeeds {
+		close(s)
+	}
+}
+
+// WaitForShutdown waits until the order processing is finished and the Market
+// is stopped.
+func (m *Market) WaitForShutdown() {
+	m.swapper.WaitForShutdown()
+	m.wg.Wait()
+}
+
+// WaitForEpochOpen waits until the start of epoch processing.
+func (m *Market) WaitForEpochOpen() {
+	<-m.running
+}
+
+func (m *Market) notifyNewOrder(order order.Order, action bookUpdateAction) {
+	// send to each receiver
+	for _, s := range m.orderFeeds {
+		select {
+		case <-m.ctx.Done():
+			return
+		case s <- &bookUpdateSignal{action, order}:
+		}
+	}
+}
+
+// Cancel the active epoch, signalling to the clients that their orders have
+// failed due to server shutdown.
+func (m *Market) cancelEpoch() {
+	//TODO
+}
```

Comment:
I put this here because I don't believe we have planned out exactly how market shutdown/halt/pause will work.  At least it's not clear in my mind.